### ISSUE TITLE
False negative `attribute-defined-outside-init` with class definition in other file

### ISSUE THREAD

**Issue by hennjanssen** [pylint-dev/pylint#9349]
### Bug description

```python
### file "mwe/__init__.py"
# empty


### file "mwe/sub_item.py"

class SubItem:
    def __init__(self, name: str = "abc") -> None:
        self.real_name: str = name


### file "mwe/item.py"

from mwe.sub_item import SubItem

class Item:
    def __init__(self) -> None:
        self.sub_item: SubItem = SubItem()

    def set_name(self, name: str) -> None:
        self.sub_item.name = name
```

### Command used

```shell
python3 -m pylint .
```

### Pylint output

```python
************* Module mwe.sub_item
sub_item.py:1:0: R0903: Too few public methods (0/2) (too-few-public-methods)
************* Module mwe.item
item.py:3:0: R0903: Too few public methods (1/2) (too-few-public-methods)
```

### Expected behavior

My expectation
----------------
```
************* Module mwe.item
item.py:8:8: W0201: Attribute 'name' defined outside __init__ (attribute-defined-outside-init)
```

Explanation
------------
In `set_name` of `class: Item` an attribute `name` is written or newly created if it does not yet exist. I would have expected to be notified by `W0201: attribute-defined-outside-init`.

The specialty here is that maybe in a previous version the attribute was `name` but was changed to `real_name` in `class: SubItem`. The user still writes to `name`, but now without effect and also without an error or pylint warning.

I suspect that it has something to do with importing the class definition and/or checking the creation of attributes in other files.

Correct behaviour
------------------
If I have both class definitions in the same file, the warning is correctly triggered.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```

### OS / Environment

- Win10 with WSL Ubuntu 22.04.3 with zsh
- but also with Win10 with git-bash
